src/libmongoc/src/mongoc/mongoc-util.c

```diff
@@ -626,3 +627,154 @@ _mongoc_getenv (const char *name)
 
 #endif
 }
+
+/* Nearly Divisionless (Algorithm 5): https://arxiv.org/abs/1805.10941 */
+static uint32_t
+_mongoc_rand_nduid32 (uint32_t s, uint32_t (*rand32) (void))
+{
+   const uint64_t limit = UINT32_MAX; /* 2^L */
+   uint64_t x, m, l;
+
+   x = rand32 ();
+   m = x * s;
+   l = m % limit;
+
+   if (l < s) {
+      const uint64_t t = (limit - s) % s;
+
+      while (l < t) {
+         x = rand32 ();
+         m = x * s;
+         l = m % limit;
+      }
+   }
+
+   return (uint32_t) (m / limit);
+}
+
+/* Java Algorithm (Algorithm 4): https://arxiv.org/abs/1805.10941
+ * The 64-bit version of the nearly divisionless algorithm requires 128-bit
+ * integer arithmetic. Instead of trying to deal with cross-platform support for
+ * `__int128`, fallback to using the Java algorithm for 64-bit instead. */
+static uint64_t
+_mongoc_rand_java64 (uint64_t s, uint64_t (*rand64) (void))
+{
+   const uint64_t limit = UINT64_MAX; /* 2^L */
+   uint64_t x, r;
+
+   x = rand64 ();
+   r = x % s;
+
+   while ((x - r) > (limit - s)) {
+      x = rand64 ();
+      r = x % s;
+   }
+
+   return r;
+}
+
+#if defined(MONGOC_ENABLE_CRYPTO)
+/* Attempt to use `_mongoc_rand_bytes` if available. */
+
+static uint32_t
+_mongoc_simple_rand_uint32_t (void)
+{
+   uint32_t res;
+
+   (void) _mongoc_rand_bytes ((uint8_t *) &res, sizeof (res));
+
+   return res;
+}
+
+static uint64_t
+_mongoc_simple_rand_uint64_t (void)
+{
+   uint64_t res;
+
+   (void) _mongoc_rand_bytes ((uint8_t *) &res, sizeof (res));
+
+   return res;
+}
+
+#else
+/* Otherwise, fallback to rudimentary use of `rand()`. */
```

Comment:
I think `rand()` needs to be preceded with a call to `srand()` to seed. Otherwise, multiple processes may select the same initial subset of hosts. From [rand(3) - Linux manual page](https://man7.org/linux/man-pages/man3/srand.3.html):

> If no seed value is provided, the rand() function is automatically seeded with a value of 1.

test-libmongoc calls `srand` from [TestSuite.c](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/tests/TestSuite.c#L656:L656). And `_bson_context_init_random` conditionally calls `srand`.

I suggest adding a call to srand in a `BSON_ONCE` function with a seed similar to the one generated in `_bson_context_init_random`.